Add support for TI TMS570LC43 SoC

ntavish opened the pull request:
The SoC has a Cortex-R5 CPU, it required big-endian and BE32 compiler flags. I enabled big-endian ARMv7-R in the SDK toolchain here, and maybe only this CPU uses it as of now: https://github.com/zephyrproject-rtos/sdk-ng/pull/765

I only have (and plan to use) only this specific SoC (TMS570LC43xx), so I focus on getting this chip working, but variants should be able to utilize this code mostly as is (AFAIK).

There is some code in the soc directory (soc.c, asm_funcs.S, soc_pll_errata.c) that is adapted from "TI Halcogen" generated code. It was license BSD 3-clause, and I am checking with TI support if they could help re-licensing it, so I did not create an RFC for the board related to that yet. I did not want to add this to the HAL module, as I am exporting only very limited portions of what the Halcogen generates, but open to suggestions.

Commit:
boards: ti: tms570_launchxl2: add new board

Adds minimal config for TI TMS570 launchpad which has the
TMS570LC43 SoC.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

Commit:
soc: ti: add TI Hercules TMS570LC43x SoC

Adds new SoC. Some of the necessary startup code is based on the
vendor provides tool (halcogen) and it includes SoC init and
some errata handling. That code originally had BSD-3-clause license,
so it cannot be part of this commit, and has been moved to TI hal
repository (modules/hal/ti).

The Halcogen code was mostly adapted to zephyr coding style and
the the dependencies were minimized.

As for erratas, the included one right now is PLL startup problem
errata.

We also have CORTEX-R5#7 (ARM ID-780125), but it was not included
because:

1. Halcogen code does not implement it
3. This is Cortex R5 errata, Xilinx R5 implementation has this note
https://adaptivesupport.amd.com/s/article/65878

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

nordicjm commented on arch/arm/core/Kconfig:
```diff
@@ -174,6 +174,14 @@ config PLATFORM_SPECIFIC_INIT
 
 	  This option is deprecated, use SOC_RESET_HOOK instead.
 
+config PLATFORM_SPECIFIC_PRE_STACK_INIT
+	bool "Platform (SOC) specific startup hook executed before the stack is initialised"
+	help
+	  The platform specific initialization code (z_arm_platform_pre_stack_init) is
+	  executed in the early startup code right after the registers are set to known
+	  values and before stack is set up. Note that this code cannot use the stack unlike
+	  z_arm_platform_init.
```
@nashif for review

nordicjm commented on soc/ti/hercules/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+	bool
+
+config SOC_FAMILY
+	default "ti_hercules" if SOC_FAMILY_HERCULES
```
names must match either: `default "hercules" if SOC_FAMILY_HERCULES` or `default "ti_hercules" if SOC_FAMILY_TI_HERCULES`, soc.yml must also match

nordicjm commented on soc/ti/hercules/tms570lc43x/Kconfig:
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_TI_TMS570LC43X
+	bool
```
```suggestion
```

nordicjm commented on soc/ti/hercules/tms570lc43x/Kconfig:
```diff
@@ -0,0 +1,21 @@
+	select PLATFORM_SPECIFIC_INIT
+	select PLATFORM_SPECIFIC_PRE_STACK_INIT
+	# select CPU_HAS_FPU
+	# select ARM_CUSTOM_INTERRUPT_CONTROLLER
```
remove if not needed

ntavish replied:
fixed

nordicjm commented on soc/ti/hercules/tms570lc43x/Kconfig:
```diff
@@ -0,0 +1,21 @@
+	# select CPU_HAS_FPU
+	# select ARM_CUSTOM_INTERRUPT_CONTROLLER
+	imply XIP
+
+if SOC_TI_TMS570LC43X
+
+endif # SOC_TI_TMS570LC43X
```
```suggestion
```

ntavish replied:
fixed

nordicjm commented on soc/ti/hercules/tms570lc43x/Kconfig.soc:
```diff
@@ -0,0 +1,20 @@
+config SOC_SERIES
+	default "ti_hercules" if SOC_SERIES_TI_HERCULES
+
+config SOC
+	default "tms570lc43x" if SOC_TI_TMS570LC43X
```
as above, needs `ti_` in string or TI_ removing from symbol, must match soc.yml

ntavish replied:
fixed

nordicjm commented on soc/ti/hercules/soc.yml:
```diff
@@ -0,0 +1,2 @@
+socs:
+  - name: tms570lc43x
```
where is family and series?

ntavish replied:
fixed

nordicjm commented on drivers/interrupt_controller/Kconfig.tms570_vim:
```diff
@@ -0,0 +1,15 @@
+	default y
+	depends on DT_HAS_TI_TMS570_VIM_ENABLED
+	help
+		The TI Vectored Interrupt Manager provides hardware assistance for prioritizing
```
help indent is 1x tab followed by 2x spaces

ntavish replied:
fixed

Commit:
drivers: pinctrl: add tms570 driver

Adds driver for TI TMS570 SoC.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

Commit:
drivers: timer: add TMS570 RTI timer as system timer

Adds the RTI timer for TI TMS570 SoC and enables it as the system timer
for enabling multithreading.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

Commit:
drivers: gpio: add TMS570 gpio driver

Adds TI TMS570 SoC GPIO driver.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

ntavish commented on dts/arm/ti/tms570.dtsi:
```diff
@@ -0,0 +1,71 @@
+		*              -> VCLK (75 MHz)
+		*                  -> RTICLK (75 MHz, divider is bypassed when VCLK is the source)
+		*/
+	};
```
@nordicjm thak you for the review, I have fixed the PR based on your comments. Sorry it was still in draft, I was rebasing and refactoring some code still from an older branch. I will still keep it in draft for today, while I add some description, and also try to figure out what to do with the BSD 3-clause code from TI.

henrikbrixandersen commented on soc/ti/hercules/ti_tms570lc43x/soc.c:
```diff
@@ -0,0 +1,334 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
We typically cannot have BSD-3-Clause licensed code in the tree, as this needs to be approved by the governing board (see https://docs.zephyrproject.org/latest/contribute/guidelines.html#components-using-other-licenses).

ntavish replied:
moved the BSD licensed code to hal_ti module (PR here https://github.com/zephyrproject-rtos/hal_ti/pull/63)

ntavish commented on boards/ti/tms570_launchxl2/tms570_launchxl2_defconfig:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_TI_TMS570=y
```
fixed

ntavish commented on soc/ti/hercules/tms570lc43x/Kconfig.defconfig:
```diff
@@ -32,10 +32,10 @@ config SYS_CLOCK_TICKS_PER_SEC
 
 config FLASH_BASE_ADDRESS
-	default 1000000
+	default 0x00000000
```
fixed

ntavish commented on arch/arm/core/Kconfig:
```diff
@@ -174,6 +174,14 @@ config PLATFORM_SPECIFIC_INIT
 
 	  This option is deprecated, use SOC_RESET_HOOK instead.
 
+config PLATFORM_SPECIFIC_PRE_STACK_INIT
```
I removed this commit and made use of soc_reset_hook after adding a small fix.

nordicjm commented on soc/ti/hercules/ti_tms570lc43x/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+config SYS_CLOCK_TICKS_PER_SEC
+	default 100
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 1000000
```
needs to come from dts property

nordicjm commented on soc/ti/hercules/ti_tms570lc43x/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 1000000
+
+config FLASH_BASE_ADDRESS
+	default 0x00000000
```
get from dts property?

ntavish replied:
fixed

nordicjm commented on soc/ti/hercules/ti_tms570lc43x/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+config FLASH_BASE_ADDRESS
+	default 0x00000000
+
+config FLASH_SIZE
+	default 4096
```
as above

ntavish replied:
fixed

Commit:
drivers: serial: add tms570 uart driver

Adds SCI peripheral UART driver for TMS570.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

Commit:
drivers: intc: add vim_tms570 interrupt controller

Add TI VIM driver specific to the TMS570 chips. This driver/peripheral
is not compatible with the existing intc_vim.c for some other TI chips.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

nordicjm commented on arch/arm/core/cortex_a_r/reset.S:
```diff
@@ -331,4 +331,5 @@ _primary_core:
 
     bl z_arm_relocate_vector_table
 
+    ldr r4, =z_prep_c
```
@wearyzen fyi

ntavish commented on dts/arm/ti/tms570.dtsi:
```diff
@@ -0,0 +1,93 @@
+				reg = <0x00000000 DT_SIZE_M(4)>;
+			};
+		};
+
```
maybe you meant line 89, removed that one.

ntavish commented on soc/ti/hercules/ti_tms570lc43x/Kconfig.soc:
```diff
@@ -0,0 +1,20 @@
+	  TI TMS570LC43X
+
+config SOC_SERIES
+	default "ti_hercules" if SOC_SERIES_TI_TMS570
```
oops, fixed

wearyzen reviewed (dismissed):
please update the soc_reset_hook to preserve r4-r7 and remove the change from reset.S

ntavish commented on soc/ti/hercules/ti_tms570lc43x/asm_funcs.S:
I do see code for ARM MPU in zephyr, including for Cortex R5 MPU, but I am unable to confirm if it does the equivalent of this, so I have to keep this halcogen generated code in for now.

Commit:
doc: boards: add tms570_launchxl2 board documentation

Adds basic information about the board about supported peripherals.

Signed-off-by: Tavish Naruka <t-naruka@ispace-inc.com>

kartben reviewed (approved):
+1 for docs, thanks

ntavish commented:
Hello @sgilbert182 I haven't updated this PR in a while. A fix is not necessary at this point, I don't plan to work further on this PR in the near future. But I do hope to pick it up again at some point of time.

The halcogen code doesn't seem to be getting merged, but I have gotten rid of much of the BSD license code, and have created more peripheral drivers. The biggest remaining piece is the Cortex-R5 MPU code, which currently is assembly exported from the halcogen GUI, but as I understand it should be compatible with standard Cortex-R MPU and cane be written in C using zephyr MPU APIs. Once I sort that out, I'll rebase to latest zephyr and update this PR (no hal-ti branch code).
